### ISSUE TITLE
Configuration in code?

### ISSUE THREAD

**Issue by Porkstone** [KentorIT/authservices#444]
First let me say that your work here has greatly simplified our Okta SAML integration so a big thank you for that.

We have a strong aversion to letting our web.config files become bloated and hard to manage. If I was to spend some time looking at adding a programmatic option to pull in the configuration data would it be worthwhile contributing this back to the project? Essentially we would just be giving the developer the choice of where to store config.

If you think this is something that would be worth contributing then please let me know if you have a view on how best to achieve this.

**Comment by AndersAbel** [comment]
What API module are you using (Owin, MVC or HttpModule)? For Owin there already is support for doing the config entirely in code without ever touching web.config. For the others, it would be interesting to explore options.

**Comment by albinsunnanbo** [comment]
The current solution for programmatic configuration is an API.
There is no available option to read a config section from a string or database, as a consumer of AuthServices you need to come up with your own storage solution for the config.

The current solution is very flexible, but requires some work.
Or maybe you can just serialize the KentorAuthServicesSection?

**Comment by heiny** [comment]
For MVC API module, is this an option?

**Comment by AndersAbel** [comment]
@heiny Is this really a question about configuration? Or do you want to know if this library can be used with a web api? In that case the answer is yes.

**Comment by heiny** [comment]
Hi. More a question if it supports programmatic configuration vs web.config for use in a multi tenancy environment, where depending on an identifier during initial login (hostnames, dns, email domain, etc), once customer is identified, the authn request, url, etc will need to be dynamically configured such that the SP sends to the customers IdP (for SP initiated SSO case).

**Comment by AndersAbel** [comment]
@heiny That's another question then. Briefly: Add all the Idps from all the customers to the config. Use the `SelectIdentityProvider` to select Idp based on the identifier you have.

If you want further assistance, I do offer commercial support and consulting, mail me at anders@sustainsys.com for more info.

**Comment by AndersAbel** [comment]
@heiny It's already possible to do programmatically as well. You can access and modify the config used at runtime through `Sustainsys.Saml2.Mvc.Saml2Controller.Options`.

**Comment by heathem** [comment]
Hello, on this topic, I want to be able to programmatically update the web.config sustainsys.saml2 section with adding an identity provider. The idea is users can use my app to enter their IdP information in a form, press "Save", and the code will add a new IdP or modify an existing one. 

I was able to use @AndersAbel suggestion and did `Sustainsys.Saml2.Mvc.Saml2Controller.Options.IdentityProviders.Add(newIdp)`. This changes the runtime version, but the actual web.config file is unchanged. I've tried using `System.Web.Configuration.WebConfigurationManager` to open the web.config file, then `Instance.GetSection("sustainsys.saml2") as SustainsysSaml2Section` to get the sustainsys.saml2 section. But all the values are readonly or have internal sets. How can update the sustainsys.saml2 section in the actual web.config file?
